File: app/src/main/java/com/itn/securebrowser/util/TimeTracker.java
package com.itn.securebrowser.util;

import android.content.Context;
import android.content.SharedPreferences;
import java.net.URI;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

public class TimeTracker {

    private final SharedPreferences prefs;
    private String trackedDomain;
    private long sessionStart;

    public TimeTracker(Context context) {
        prefs = context.getSharedPreferences("itn_time_tracker", Context.MODE_PRIVATE);
    }

    public void onDomainChanged(String newDomain) {
        commitCurrentSession();
        trackedDomain = newDomain;
        sessionStart = (newDomain != null) ? now() : 0L;
    }

    public void onAppPaused() {
        commitCurrentSession();
    }

    public void onAppResumed(String domain) {
        trackedDomain = domain;
        sessionStart = (domain != null) ? now() : 0L;
    }

    public long getTodaySeconds(String domain) {
        long stored = prefs.getLong(key(domain), 0L);
        long live = (domain.equals(trackedDomain) && sessionStart > 0L) ? now() - sessionStart : 0L;
        return stored + live;
    }

    private void commitCurrentSession() {
        if (trackedDomain == null || sessionStart <= 0L) return;
        long elapsed = now() - sessionStart;
        if (elapsed <= 0L) return;
        String k = key(trackedDomain);
        prefs.edit().putLong(k, prefs.getLong(k, 0L) + elapsed).apply();
        sessionStart = 0L;
        trackedDomain = null;
    }

    private String key(String domain) {
        return domain + "_" + todayString();
    }

    private String todayString() {
        return new SimpleDateFormat("yyyy-MM-dd", Locale.US).format(new Date());
    }

    private long now() {
        return System.currentTimeMillis() / 1000L;
    }

    public static String extractDomain(String url) {
        if (url == null || url.trim().isEmpty() || "about:blank".equals(url)) return null;
        try {
            String host = URI.create(url).getHost();
            if (host == null || host.isEmpty()) return null;
            if (host.startsWith("www.")) host = host.substring(4);
            return host.isEmpty() ? null : host;
        } catch (Exception e) {
            return null;
        }
    }
}

File: app/src/main/java/com/itn/securebrowser/TimeTracker.kt
package com.itn.securebrowser

import android.content.Context
import android.content.SharedPreferences
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class TimeTracker(context: Context) {

    private val prefs: SharedPreferences =
        context.getSharedPreferences("itn_time_tracker", Context.MODE_PRIVATE)

    private var trackedDomain: String? = null
    private var sessionStart: Long = 0L

    // ── Public API ─────────────────────────────────────────────────────────

    /** استدعه عند تغيّر النطاق النشط: صفحة جديدة، تبديل تبويب، أو لا شيء. */
    fun onDomainChanged(newDomain: String?) {
        commitCurrentSession()
        trackedDomain = newDomain
        sessionStart = if (newDomain != null) now() else 0L
    }

    /** استدعه في onPause — يحفظ الجلسة الحالية. */
    fun onAppPaused() {
        commitCurrentSession()
    }

    /** استدعه في onResume مع النطاق المرئي حالياً. */
    fun onAppResumed(domain: String?) {
        trackedDomain = domain
        sessionStart = if (domain != null) now() else 0L
    }

    /**
     * إجمالي الثواني المسجّلة اليوم لـ [domain]،
     * يشمل الجلسة الجارية حالياً إن وُجدت.
     */
    fun getTodaySeconds(domain: String): Long {
        val stored = prefs.getLong(key(domain), 0L)
        val live = if (domain == trackedDomain && sessionStart > 0L)
            now() - sessionStart else 0L
        return stored + live
    }

    /** جميع النطاقات التي تم تتبّعها اليوم مع إجمالي ثوانيها (المحفوظة فقط). */
    fun getAllTodayData(): Map<String, Long> {
        val suffix = "_${todayString()}"
        return prefs.all
            .filter  { (k, _) -> k.endsWith(suffix) }
            .mapKeys { (k, _) -> k.removeSuffix(suffix) }
            .mapValues { (_, v) -> (v as? Long) ?: 0L }
    }

    // ── Internal ───────────────────────────────────────────────────────────

    private fun commitCurrentSession() {
        val domain = trackedDomain ?: return
        if (sessionStart <= 0L) return
        val elapsed = now() - sessionStart
        if (elapsed <= 0L) return
        val k = key(domain)
        prefs.edit().putLong(k, prefs.getLong(k, 0L) + elapsed).apply()
        sessionStart = 0L
        trackedDomain = null
    }

    private fun key(domain: String) = "${domain}_${todayString()}"

    private fun todayString(): String =
        SimpleDateFormat("yyyy-MM-dd", Locale.US).format(Date())

    private fun now() = System.currentTimeMillis() / 1000L

    companion object {
        /** يستخرج النطاق (بدون www.) من URL، أو null إن لم يكن قابلاً للتتبع. */
        fun extractDomain(url: String?): String? {
            if (url.isNullOrBlank() || url == "about:blank") return null
            return try {
                val host = java.net.URI(url).host ?: return null
                host.removePrefix("www.").takeIf { it.isNotBlank() }
            } catch (e: Exception) {
                null
            }
        }
    }
}
